src/framework/operations.hpp

```diff
@@ -571,6 +574,44 @@ Op json_to_op_initialize(const json_t &js) {
   return op;
 }
 
+Op json_to_op_initialize_ket(const json_t &js) {
+  // creates a standard statevector-based initialize op
+  // from the input ket; faster than converting in Python
+  Op op;
+  op.type = OpType::initialize;
+  op.name = "initialize";
+  JSON::get_value(op.qubits, "qubits", js);
+
+  uint_t nint = op.qubits.size();
+
+  std::map<std::string, complex_t> ketmap;
+
+  if (JSON::check_key("params", js) && js["params"].is_array()) {
+    for (const auto &comp : js["params"]) {
+      // Check component is length-2 array
+      if (!comp.is_array() || comp.size() != 2)
+        throw std::invalid_argument("Invalid initialize_ket (param component " +
+                                    comp.dump() + " invalid).");
+      // Get amplitude string
+      std::string key = comp[0];
+      // Get complex coefficient
+      complex_t coeff = comp[1];
+      ketmap[key] = coeff;
+    }
+  }
+
+  op.params = Utils::ket2vec(ketmap, nint, 16);
```

Comment:
Instead of converting to a dense vector here we could be more efficient by defining a new OpType `initialize_ket`, and storing just the non-zero coefficients and the index positions of the string labels in `params` and `regs` (or `string_params`) respectively: ie something like:

`{'00': 0.707.. , '11': 1j * 0.707}` -> `params = [0.707..., 1j * 0.707...], regs = [[0, 3]]`

Note `regs` here isn't a perfect field for storing this (it stores vectors of ints) so we just put all the coeffs as the first entry of `regs`. We could have used string params and stored the bit-strings but that would just delay conversion of strings to ints later which would be less efficient if being called over multiple shots.

To initialize a vector using this in the state class we can then initialize an empty vector/matrix, and then iterate over params/regs[0] and update the corresponding vector entries.

Implementing this might involve a bigger change to how general initialization is handled in the controller class so maybe we can leave this as parsing to a regular initialize instruciton now, and implement this ket OpType as a followup (and while doing so fix initialization for the non-statevector simulators to work properly)

Reply:
We're getting the coefficients/index values already in this suggested format (via `js["params"]`) and can indeed use it in a new initialization op as you suggested; but I think it's best to delay this to the future, again as you suggested.